File: app/src/main/java/com/marvelcomics/brito/marvelcomics/ui/fragment/comics/ComicsFragment.java
package com.marvelcomics.brito.marvelcomics.ui.fragment.comics;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.marvelcomics.brito.entity.ComicEntity;
import com.marvelcomics.brito.infrastructure.utils.AlertDialogUtils;
import com.marvelcomics.brito.marvelcomics.R;
import com.marvelcomics.brito.marvelcomics.databinding.FragmentComicsBinding;
import com.marvelcomics.brito.marvelcomics.ui.fragment.ItemOffSetDecorationHorizontal;
import com.marvelcomics.brito.presentation.presenter.comics.ComicsContract;
import com.marvelcomics.brito.presentation.presenter.comics.ComicsPresenter;

import java.util.List;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

public class ComicsFragment extends Fragment implements ComicsContract.View {

    private static final String ARGUMENT_CHARACTER_ID = "character_id_args";
    private int characterId;

    @Inject
    protected ComicsPresenter comicsPresenter;

    private FragmentComicsBinding binding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        characterId = args.getInt(ARGUMENT_CHARACTER_ID);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        AndroidSupportInjection.inject(this);
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_comics, container, false);
        comicsPresenter.setView(this);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        loadComics();
    }

    @Override
    public void showComics(List<ComicEntity> comics) {
        createdAdapter(comics);
    }

    @Override
    public void showError(String message) {
        AlertDialogUtils.showSimpleDialog("Erro", message, getContext());
    }

    private void loadComics() {
        comicsPresenter.loadComics(characterId);
    }

    private void createdAdapter(List<ComicEntity> comicEntities) {
        ComicsAdapter adapter = new ComicsAdapter(comicEntities);
        binding.recyclerviewFragmentComic.setLayoutManager(
                new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        binding.recyclerviewFragmentComic.setAdapter(adapter);
        binding.recyclerviewFragmentComic.addItemDecoration(new ItemOffSetDecorationHorizontal(8));
    }

    public static ComicsFragment newInstance(int characterId) {
        ComicsFragment fragment = new ComicsFragment();

        Bundle args = new Bundle();
        args.putInt(ARGUMENT_CHARACTER_ID, characterId);

        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/com/marvelcomics/brito/view/fragment/comics/ComicsFragment.kt
package com.marvelcomics.brito.view.fragment.comics

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.marvelcomics.brito.R
import com.marvelcomics.brito.databinding.FragmentComicsBinding
import com.marvelcomics.brito.domain.entity.ComicEntity
import com.marvelcomics.brito.infrastructure.utils.AlertDialogUtils
import com.marvelcomics.brito.view.extensions.viewBinding
import com.marvelcomics.brito.view.fragment.ItemOffSetDecorationHorizontal
import com.marvelcomics.brito.viewmodel.ComicUiState
import com.marvelcomics.brito.viewmodel.GlobalUiState
import com.marvelcomics.brito.viewmodel.comic.ComicViewModel
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.flow.collect
import org.koin.android.ext.android.inject

@InternalCoroutinesApi
class ComicsFragment : Fragment(R.layout.fragment_comics) {

    private val binding by viewBinding(FragmentComicsBinding::bind)
    private val comicViewModel: ComicViewModel by inject()

    private var characterId: Int? = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        characterId = arguments?.getInt(ARGUMENT_CHARACTER_ID)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initObservers()
        loadComics()
    }

    private fun initObservers() {
        lifecycleScope.launchWhenStarted {
            comicViewModel.comicUiState.collect {
                when (it) {
                    is ComicUiState.Success -> {
                        showComics(it.data as List<ComicEntity>)
                    }
                    is ComicUiState.Error -> {
                        it.exception.message?.let { message ->
                            showError(message)
                        }
                    }
                    is GlobalUiState.Loading -> {
                        showLoading()
                    }
                    is GlobalUiState.NetworkError -> {
                        // do nothing
                    }
                    else -> {
                        // do nothing
                    }
                }
            }
        }
    }

    private fun loadComics() {
        comicViewModel.loadComics(characterId ?: let { 0 })
    }

    private fun showComics(comics: List<ComicEntity>) {
        binding.progressbarLoadingComics.visibility = View.GONE
        binding.recyclerviewFragmentComic.visibility = View.VISIBLE
        createAdapter(comics)
    }

    private fun showLoading() {
        binding.progressbarLoadingComics.visibility = View.VISIBLE
        binding.recyclerviewFragmentComic.visibility = View.GONE
    }

    private fun showError(message: String) {
        binding.progressbarLoadingComics.visibility = View.GONE
        binding.recyclerviewFragmentComic.visibility = View.VISIBLE
        AlertDialogUtils.showSimpleDialog("Erro", message, requireContext())
    }

    private fun createAdapter(listComics: List<ComicEntity>) {
        val comicsAdapter = ComicsAdapter(listComics)
        binding.recyclerviewFragmentComic.layoutManager =
            LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        binding.recyclerviewFragmentComic.adapter = comicsAdapter
        binding.recyclerviewFragmentComic.addItemDecoration(
            ItemOffSetDecorationHorizontal(8)
        )
    }

    companion object {
        const val ARGUMENT_CHARACTER_ID = "character_id_args"

        fun newInstance(characterId: Int): ComicsFragment {
            val fragment = ComicsFragment()
            val args = Bundle()
            args.putInt(ARGUMENT_CHARACTER_ID, characterId)

            fragment.arguments = args
            return fragment
        }
    }
}
